### ISSUE TITLE
Typographical quotes should respect user language

### ISSUE THREAD

**Issue by spixi** [diaspora/diaspora#7439]
# Symptome
When I insert double and single quotes in d*  I normally use the replacement characters \" and \'. After saving the comment, d* replaces those characters by typographical quotes, like in “my text”. However this is done independently of the user language.

# Expected Behaviour
When I set my d* language to another language  I expect to receive „German“,  « French » or 「Ｊａｐａｎｅｓｅ」quotes, instead.

# Actual Behaviour
Setting the d* language to German or any other language produces “US american” quotes.

**Comment by spixi** [comment]
![the example shows US American quotes erroneously set  in German text](https://cloud.githubusercontent.com/assets/3192596/25639765/5ed9be18-2f8d-11e7-9efb-a4e31b8f3e48.png)

**Comment by svbergerem** [comment]
These quotes would have to depend on the language of the text instead of the user's chosen language. Here's an example how you can set quote replacements in markdown-it: https://github.com/markdown-it/markdown-it#init-with-presets-and-options

**Comment by spixi** [comment]
Good point, @svbergerem. But this would require a language detection heuristic and this would add much overhead (I know what I am telling about, because I have already wrote such an algorithm based using hidden markov models). A very trivial script determination heuristic could detect, e. g. Greek, Hebrew or Georgian texts, but would be unable do distinguish between languages using the same script (like Latin or Cyrillic) As mentioned above, I think it is better to rely on the user’s language, because the probability that a person with a French UI language usually writes French texts is much higher than usually writing Spanish texts, for example, so the UI language is often, but not always the best guess. If some other quots are required, the user has to set the quotes themselves or escape the quote replacement characters with the back slash to bypass wuote conversion.

Maybe we should add fast-access keys for those characters, like in MediaWiki's text editor widget, if typographic quotes are such important (_Off-topic: I also would like to have a button for the "zero-with non-breaking space" character which is usually to terminate hash tags without breaking a word_)

An alternative suggestion would be: Don't convert the replacement quotes at all, because they may also appear in some units of measures like degrees and inches. If a user needs typographic quotes he would have to intentionally set them.

What do others think?

**Comment by Flaburgan** [comment]
>What do others think?

I have diaspora set in french but mostly write in english. I know people who always use the software in english but often write in french. I don't think that we should base ourselves on the user language.

So the only way to correctly do that would be to detect the language of the post. This is a highly asked feature to be able to filter the stream depending of language posts btw.

**Comment by goobertron** [comment]
If you are writing in another language, wouldn't you have your keyboard set to enter the quote marks for that language rather than "..." or '...'?

**Comment by Waithamai** [comment]
for me it's the same as for @Flaburgan. my diaspora is english, although i write mostly in german. but I also write in french and in spanish, soo.. nope, using diaspora's user settings would definitely not work. 

also ++ for goob, especially for languages that use other character sets than latin.

**Comment by spixi** [comment]
As mentioned above, if language detection is desired I could try this from scratch. I remember a hidden markov model implemention which simply counts the occurance of adjacent-letter pairs and matches the result against predetermined language-specific tables. That implementation was ASCII-only, but this is too much overhead imho.

Sure, you may work with a custom keyboard layout for writing, but this is neither comfortable, nor portable.

(I am only familar with T2 (QWERTZ), which actually *has* German quotes, but this requires to make use of the annoying group switch key (in my case AltGr+^), which is actually used very rare. Using the replacement character works abosutely fine, because the text it is parsed and displayed, as if you had send typographical quotes. This is only true for English. I hope, you understand my problem.

**Comment by spixi** [comment]
The current implementation is just not optimal and violates the principle of least surprise. When I write a German text, I don't expect, that the simple upright quotes are replaced by proper US american quotes. This is more incorrect than simply keeping the upright quotes like they are.

**Comment by tenno-seremel** [comment]
IMO, the best way is to turn automation off and (maybe) add buttons with different quotes. Although they are kind of numerous: ‘’‚‛“”„‟«» etc.

// Random diaspora user.

**Comment by spixi** [comment]
@tenno-seremel +1
